### PR TITLE
[PropertyInfo] Add Component Documentation

### PR DESCRIPTION
**Pull request by zanbaldwin**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes (symfony/symfony#15858) |
| Applies to | `2.8`, `3.0` |
| Fixed tickets | N/A |

This is still a _work in progress_, but I'd like feedback from anyone regarding structure, grammar and any other improvements while I continue. After speaking with @dunglas, I'm creating a new PR due our branches differing by about 400 commits.

### CODE REVIEW

File: components/property_info.rst [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+==========================
+
+The PropertyInfo component provides the functionality to get information about class properties using metadata of
+popular sources.
+
+Whilst the :doc:`PropertyAccess component</components/property_access/introduction>` allows you to read and write values
+to/from objects and arrays, the PropertyInfo component works solely with class definitions to provide information
+such as data type and visibility about properties within that class.
+
+Similar to PropertyAccess, the PropertyInfo component blurs the lines between class properties such as ``$property``
```
Comment (zanderbaldwin) [inline components/property_info.rst]: As pointed out by a friend in the Philippines, the idiom *blurs the lines* is not easily understood by non-native English speakers. Need to consider an alternative.

File: components/property_info.rst [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+and properties defined via accessor and mutator methods such as  ``getProperty()``, ``isProperty()``, ``setProperty()``,
```
Comment (dunglas) [inline components/property_info.rst]: It is also able to leverage adders and removers.

File: components/property_info.rst [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    use Symfony\Component\PropertyInfo\PropertyInfoExtractor;
+
+    $propertyInfo = new PropertyInfoExtractor(
+        $arrayOfListExtractors,
+        $arrayOfTypeExtractors,
+        $arrayOfDescriptionExtractors,
+        $arrayOfAccessExtractors
+    );
+
+The order of extractor instances within an array matter, as the first non-null result will be returned.
```
Comment (dunglas) [inline components/property_info.rst]: matters?

File: components/property_info.rst [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+That is why you must provide each category of extractors as a separate array, even if an extractor provides information
+for more than one category.
+
+For example, whilst the :class:`ReflectionExtractor<Symfony\\Component\\PropertyInfo\\Extractor\\ReflectionExtractor>`
+and :class:`DoctrineExtractor<Symfony\\Bridge\\Doctrine\\PropertyInfo\\DoctrineExtractor>` both provide list and type
+information it is probably best that:
```
Comment (dunglas) [inline components/property_info.rst]: better?

File: components/property_info.rst [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+.. _`components-property-info-type`:
+
+Type Objects
+------------
+
+Compared to the other extractors, type information extractors provide much more information than can be represented as
+simple scalar values - because of this, type extractors return an array of objects. The array will contain an instance
+of the :class:`Type<Symfony\\Component\\PropertyInfo\\Type>` class for every parameter of a property's mutator method.
+
+.. If you're getting the "type" of a property, should an array of types be returned? If a mutator requires more than
```
Comment (dunglas) [inline components/property_info.rst]: * An array of types is returned to support properties with several types like `int|float`
* Yes, only mutators with one **required** parameter are taken into account and IIRC it's the type of the required parameter that is used

**Comment by dunglas** [comment]
Good job! I left some minor comments but :+1:

**Comment by zanderbaldwin** [comment]
Considering I failed English at school, the only language I can speak, thanks!
I'll try to finish this along with the edits within the next couple of days.

File: components/property_info.rst (hunk 3 above)
Comment (dupuchba) [inline components/property_info.rst]: I understood : "The order of extractor instance is important"

**Comment by zanderbaldwin** [comment]
Back in Europe, finally have internet access, any comments on this so far? /cc @dunglas.

**Comment by TomasVotruba** [comment]
This looks good to me. Anything more that needs to be done?

**Comment by zanderbaldwin** [comment]
The *Creating Your Own Extractors* section could possibly be expanded but I could never figure out how to word it :P

**Comment by TomasVotruba** [comment]
I think that might be next, advanced article.

I'd add just this simple "How to use this" version. It's really missed at the moment.

File: components/property_info.rst (hunk 1 above)
Comment (teohhanhui) [inline components/property_info.rst]: What about "works with both"?

File: components/property_info.rst (hunk 2 above)
Comment (teohhanhui) [inline components/property_info.rst]: This list of example method names is not illustrative. I think some actual examples would be better.

File: components/property_info.rst [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    single: PropertyInfo
+    single: Components; PropertyInfo
+
+The PropertyInfo Component
+==========================
+
+The PropertyInfo component provides the functionality to get information about class properties using metadata of
+popular sources.
+
+Whilst the :doc:`PropertyAccess component</components/property_access/introduction>` allows you to read and write values
```
Comment (teohhanhui) [inline components/property_info.rst]: I see multiple occurrences of "whilst".

See http://dictionary.cambridge.org/grammar/british-grammar/while-and-whilst

Comment (zanderbaldwin) [inline components/property_info.rst]: I'll edit that now.

File: components/property_info.rst [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+==========================
+
+The PropertyInfo component provides the functionality to get information about class properties using metadata of
+popular sources.
+
+While the :doc:`PropertyAccess component</components/property_access/introduction>` allows you to read and write values
+to/from objects and arrays, the PropertyInfo component works solely with class definitions to provide information
+such as data type and visibility about properties within that class.
+
+Similar to PropertyAccess, the PropertyInfo component blurs the lines between class properties such as ``$property``
```
Comment (teohhanhui) [inline components/property_info.rst]: Previous discussion on rewording of "blurs the lines between" here: https://github.com/symfony/symfony-docs/pull/5974#discussion_r46890491

Perhaps "combines both"?

File: components/property_info.rst [hunk 8]
```diff
@@ -0,0 +1,432 @@
+.. index::
+    single: PropertyInfo
+    single: Components; PropertyInfo
+
+The PropertyInfo Component
+==========================
+
+The PropertyInfo component provides the functionality to get information about class properties using metadata of
+popular sources.
```
Comment (WouterJ) [inline components/property_info.rst]: this paragraph should be indented with 4 spaces to be consistent with other component docs.

File: components/property_info.rst [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+While the :doc:`PropertyAccess component</components/property_access/introduction>` allows you to read and write values
```
Comment (WouterJ) [inline components/property_info.rst]: very minor, but there should be a space before `<`

File: components/property_info.rst [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+Additional dependencies may be required for some of the
+:ref:`extractors provided with this component<components-property-info-extractors>`.
+
+.. _`components-property-information-usage`:
+
+Usage
+-----
+
+The entry point of this component is a new instance of the
+:class:`PropertyInfoExtractor<Symfony\\Component\\PropertyInfo\\PropertyInfoExtractor>` class, providing sets of
```
Comment (WouterJ) [inline components/property_info.rst]: the class role can be: ``:class:`Symfony\\Component\\PropertyInfo\\PropertyInfoExtractor` ``, no need for a label as the label matches the default label.

File: components/property_info.rst [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+        // Type extractors
+        array(
+            $doctrineExtractor,
+            $reflectionExtractor
+        )
+    );
+
+.. _`components-property-information-usage-configuration`:
+
+Configuration
```
Comment (WouterJ) [inline components/property_info.rst]: the component docs shouldn't document framework integration. This has to be added to the config reference

**Comment by WouterJ** [comment]
Hi @zanderbaldwin!

I'm sorry that it took so long before one of us reviewed this PR. I think it's a nice start. Can you please fix the comments I added to this PR and make sure lines are wrapped after the first word that crosses the 72th character (they are a bit too long atm)? Thanks already!

If you don't have time, just comment and we'll continue this PR.

status: needs work

**Comment by zanderbaldwin** [comment]
Hi guys, I've made the changes that were suggested. Does anyone have any other input?

**Comment by dunglas** [comment]
Looks good to me. You did a very good job documenting this component, thank you!

**Comment by dunglas** [comment]
@zanderbaldwin can you rebase your PR?

@xabbuh @wouterj @weaverryan can this PR be merged? It would be nice to have the doc for this component online.

**Comment by zanderbaldwin** [comment]
Rebased, though you might want to check that I've solved all the conflicts correctly...

File: components/property_info.rst [hunk 12]
```diff
@@ -1,26 +1,446 @@
...
 
 The PropertyInfo Component
 ==========================
 
-    The PropertyInfo component extracts information about PHP class properties
-    using metadata from popular sources, like `Reflection`_ and `PHP Documentation`_.
+    The PropertyInfo component provides the functionality to get information
+    about class properties using metadata of popular sources.
+
+While the :doc:`PropertyAccess component </components/property_access/introduction>`
```
Comment (dunglas) [inline components/property_info.rst]: The path should be updated to `/components/property_info` (it's why Travis fails)

Comment (teohhanhui) [inline components/property_info.rst]: * `/components/property_access`

**Comment by zanderbaldwin** [comment]
Got the build working, would you like the 5 commits squashed into 1?

**Comment by dunglas** [comment]
Commits can be squashed by the merged.

Status: reviewed

**Comment by weaverryan** [comment]
Merged! This is great! I was very much looking forward to the docs for this, but bootstrapping the docs for a new component is a lot of work :). So, thank you!

I opened #6896 with some proofreading tweaks that I'm making. I'll finish it shortly, but I just got interrupted :).

Cheers!

**Pull request merged**